File: leetcode/problems/go/1833. Maximum Ice Cream Bars/solve.go
package main

func maxIceCream(costs []int, coins int) int {
	maxCost := 0
	for _, cost := range costs {
		if cost > maxCost {
			maxCost = cost
		}
	}

	counts := make([]int, maxCost+1)
	for _, cost := range costs {
		counts[cost]++
	}

	amount := 0
	for cost, count := range counts {
		if coins < cost {
			break
		} else if count == 0 {
			continue
		}
		buyCount := min(coins/cost, count)
		amount += buyCount
		coins -= cost * buyCount
	}

	return amount
}

File: leetcode/problems/python3/1833. Maximum Ice Cream Bars/solve.py
from typing import List


class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        counts = [0] * (max(costs) + 1)
        for cost in costs:
            counts[cost] += 1

        amount = 0
        for cost, count in enumerate(counts):
            if coins < cost:
                break
            elif count == 0:
                continue
            buy_count = min(coins // cost, count)
            amount += buy_count
            coins -= cost * buy_count

        return amount
